src/rewrite/vct/rewrite/lang/LangLLVMToCol.scala

```diff
@@ -2,57 +2,305 @@ package vct.rewrite.lang
+
+  import LangLLVMToCol._
+
   type Post = Rewritten[Pre]
   implicit val implicitRewriter: AbstractRewriter[Pre, Post] = rw
 
   private val llvmFunctionMap
-      : SuccessionMap[LlvmFunctionDefinition[Pre], Procedure[Post]] =
+      : SuccessionMap[LLVMFunctionDefinition[Pre], Procedure[Post]] =
     SuccessionMap()
   private val specFunctionMap
-      : SuccessionMap[LlvmSpecFunction[Pre], Function[Post]] = SuccessionMap()
+      : SuccessionMap[LLVMSpecFunction[Pre], Function[Post]] = SuccessionMap()
+  private val globalVariableMap
+      : SuccessionMap[LLVMGlobalVariable[Pre], HeapVariable[Post]] =
+    SuccessionMap()
+  private val structMap: SuccessionMap[LLVMTStruct[Pre], Class[Post]] =
+    SuccessionMap()
+  private val structFieldMap
+      : SuccessionMap[(LLVMTStruct[Pre], Int), InstanceField[Post]] =
+    SuccessionMap()
+
+  private val globalVariableInferredType
+      : mutable.HashMap[LLVMGlobalVariable[Pre], Type[Pre]] = mutable.HashMap()
+  private val localVariableInferredType
+      : mutable.HashMap[Variable[Pre], Type[Pre]] = mutable.HashMap()
+  private val loopBlocks: mutable.ArrayBuffer[LLVMBasicBlock[Pre]] = mutable
+    .ArrayBuffer()
+  private val elidedBackEdges: mutable.Set[LabelDecl[Pre]] = mutable.Set()
+
+  def gatherBackEdges(program: Program[Pre]): Unit = {
+    program.collect { case loop: LLVMLoop[Pre] =>
+      elidedBackEdges.add(loop.header.decl)
+    }
+  }
 
-  def rewriteLocal(local: LlvmLocal[Pre]): Expr[Post] = {
+  def gatherTypeHints(program: Program[Pre]): Unit = {
+    // TODO: We also need to do something where we only keep structurally distinct types
+    def moreSpecific(self: Type[Pre], other: Type[Pre]): Boolean = {
+      (self, other) match {
+        case (a, b) if a == b => false
+        case (LLVMTPointer(None), _) => false
+        case (_, LLVMTPointer(None)) => true
+        case (LLVMTPointer(Some(a)), LLVMTPointer(Some(b))) =>
+          moreSpecific(a, b)
+        case (LLVMTPointer(Some(a)), TPointer(b)) => moreSpecific(a, b)
+        case (TPointer(a), LLVMTPointer(Some(b))) => moreSpecific(a, b)
+        case (TPointer(a), TPointer(b)) => moreSpecific(a, b)
+        case (LLVMTStruct(_, _, a), LLVMTStruct(_, _, b)) =>
+          a.headOption.exists(ta => b.exists(tb => moreSpecific(ta, tb)))
+        case (LLVMTStruct(_, _, _), _) => true
+        case (LLVMTArray(_, a), LLVMTArray(_, b)) => moreSpecific(a, b)
+        case (LLVMTArray(_, _), _) => true
+        case _ => false
+      }
+    }
+
+    // TODO: This sorting is non-stable which might cause nondeterministic bugs if there's something wrong with moreSpecific
+    def findMostSpecific(
+        types: mutable.ArrayBuffer[Type[Pre]]
+    ): Option[Type[Pre]] = {
+      types.map(Some(_)).reduce[Option[Type[Pre]]] { (a, b) =>
+        (a, b) match {
+          case (None, _) | (_, None) => None
+          case (Some(a), Some(b))
+              if a == b || rw.dispatch(a) == rw.dispatch(b) ||
```

Comment:
Comparing things in `Post` is highly dubious! Might trigger comparison of some uninitialized refs. I know sycl used to do this every once in a while and it's a pain.

Is it maybe possible to make a more explicit ordering for a fragment of types? As this is for llvm, you only need to do this for say ints, structs, and the spec types (seq/map etc.). And for the spec types you can maybe just say to not let them interleave with llvm types maybe, to simplify things? Relying on the behaviour of another rewriter is fragile. It's also fine if your ordering is very limited and needs to be extended every once in a while. What about only using `moreSpecific`, and maybe making `moreSpecific` more complete for the identity cases?

Reply:
Yeah I knew this was dubious when I wrote it. I do it because I'm mixing LLVM and PVL in places and their types will only match after the LangSpecificToCol pass has finished. This was mostly a cop-out for not having to write out every LLVM type that might match a PVL type and vice-versa.